### ISSUE TITLE
[BUG]: DynamicVector.__setitem__() doesn't raise an error when index is out of bounds

### ISSUE THREAD

**Issue by AlaDeli7** [modularml/mojo#1425]
### Bug description

Error when adding a String element to a DynamicVector

### Steps to reproduce

```
fn main():
    var v = DynamicVector[String](6)
    for i in range(4):
        v[i] = "Hello"
```

![image](https://github.com/modularml/mojo/assets/6042311/23385d91-de98-4853-8967-5600ddeb90cd)

### System information

```shell
- What OS did you do install Mojo on ?
wsl on windows 10
- Provide version information for Mojo by pasting the output of `mojo -v`
mojo 0.6.0 (d55c0025)
- Provide Modular CLI version by pasting the output of `modular -v`
modular 0.2.2 (95d42445)
```

**Comment by arthurevans** [comment]
I can't reproduce the crash on my system (macOS), but the code doesn't work as written. On my system, it fails silently, which is not ideal, either. 

Basically, `v[i] = "Hello"` replaces the value at `v[i]`. It fails if there isn't an element at `v[i]` already. This is basically the same as Python, except that Python produces an error:

```
>>> v = []
>>> v[0]="Hello"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list assignment index out of range
```

Replace `v[i] = "Hello"` with `v.append("Hello")` and it should work fine.

I'm going to retitle this issue because I think `DynamicVector` should raise an error in this case.

**Comment by JoeLoser** [comment]
Internally, we do raise an assertion if the index is not within bounds for certain build configurations.  Right now, the mojo-sdk is only shipped without assertions enabled in the Standard Library.  Given that I've now seen numerous issues around how Mojo crashes instead of an assertion firing, I think we should include this in the docs somewhere @arthurevans as it is indeed a sharp edge.  We do have an internal issue open that I filed that we do plan on addressing for allowing users to get these safety assertions in the standard library; the details are still TBD, but it's something we definitely all want.

In the short-term, I'll probably create an external issue just so we can consolidate the several issues that are open today where we _already have asserts_ in the standard library.  This external issue will mostly be for visibility sake until we address the internal one.

**Comment by JoeLoser** [comment]
I decided to implement something in the short-term for external users to opt-in to the assertions that exist in the standard library.  This just landed internally and will be available in the 0.7.0 release in January.  Assertions in the standard library will still be off/disabled by default for performance reasons.  But, to opt in for all of the assertions (including bounds checking, like in your case), you can add `-D MOJO_ENABLE_ASSERTIONS` to your `mojo` invocation.  For example,

```
mojo -D MOJO_ENABLE_ASSERTIONS my_mojo_source_file.mojo
```

would be the way to invoke `mojo` and opt in for the assertions in the standard library.  The entry in the changelog will give all the relevant details, but this is the gist of it. 😄

**Comment by JoeLoser** [comment]
In your example:

```
from collections.vector import DynamicVector

fn main():
    var v = DynamicVector[String](6)
    for i in range(4):
        v[i] = "Hello"
```

If we compile this file with `-D MOJO_ENABLE_ASSERTIONS`, we get the following error before the stack trace:

```
mojo -D MOJO_ENABLE_ASSERTIONS repro.mojo
Assert Error: end must be greater than or equal to start when step is 1
```

This assertion error comes from `range.mojo` and it could definitely be improved to state the start and end indices in question that it's firing from when trying to form the range `[start;end]`.  I think we'll have quite a few QoI-related issues with assertion messages that we'll clean up along the way of things.  When you try it out in the January release (0.7.0), please do file any QoI related issues you see with poor assertion messages and we'll fix them to make your experience better.

Thanks again for the bug report; given that this is resolved in the 0.7.0 release, I'm going to go ahead and close this for now.
